lib/active_merchant/billing/gateways/blue_snap.rb

```diff
@@ -380,5 +457,50 @@ def forbidden_response(body)
         { 'description' => body }
       end
     end
+
+    class PaymentMethodDetails
```

Comment:
Whoa, this is pretty heavy duty. I don't think any other adapters define internal classes for something like this; a `check` payment method class already exists. How necessary is this?

Reply:
For reference, here is a typical ACH implementation https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/card_connect.rb#L192-L213

Reply:
@curiousepic There is one other place in AM that defines an internal class. I started off keeping track off the payment method details using a hash, however, this quickly became unruly as I had to perform checks multiple places to see if the payment method is an alt transaction and/or if it is a check. For this particular case, creating a class with accessors allowed for better encapsulation of functionality.

Unlike a [straightforward implementation](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/card_connect.rb#L192-L213), Blue Snap's API changes the resource URL, XML root element and other aspects based on if the transaction is a card or an ACH/SEPA/Paypal/etc (`alt transaction`).  What we end up with is needing to classify 4 payment method types 1) String 2) String, but storing an Alt payment method (ACH) 3) Check, 4) [Default] Credit card.

My first pass required values to be passed repeatedly to different methods - because `url` and `root_element` are deeply nested in `commit` I had to pass data from method to method to end up with `if-elsif-else` statements replacing the ternary form `action == :store ? 'vaulted-shopper' : 'card-transaction'`. 

I'd prefer to not use an internal class but felt it was much cleaner to do so because of the encapsulation. I'm happy to pair and walk through these changes.

Reply:
I seeeee, got it. Just wanted to make sure it was justified and sounds like it is. Thanks for the details.